src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -358,9 +328,13 @@ object JoinIndexRule
     require(resolve(spark, lRequiredIndexedCols, lRequiredAllCols).isDefined)
     require(resolve(spark, rRequiredIndexedCols, rRequiredAllCols).isDefined)
 
-    val lUsable = getUsableIndexes(lIndexes, lRequiredIndexedCols, lRequiredAllCols)
-    val rUsable = getUsableIndexes(rIndexes, rRequiredIndexedCols, rRequiredAllCols)
-    val compatibleIndexPairs = getCompatibleIndexPairs(lUsable, rUsable, lRMap)
+    val lUsable = getUsableIndexes(allIndexes, lRequiredIndexedCols, lRequiredAllCols)
+    val rUsable = getUsableIndexes(allIndexes, rRequiredIndexedCols, rRequiredAllCols)
+
+    val lIndexes = RuleUtils.getCandidateIndexes(lUsable, RuleUtils.getLogicalRelation(left).get)
+    val rIndexes = RuleUtils.getCandidateIndexes(rUsable, RuleUtils.getLogicalRelation(right).get)
```

Comment:
ditto: (Similar to comment on FilterIndex rule) Please add a brief comment on why schema check should be done before signature computation/comparison.